src/main/flight/mixer.c

```diff
@@ -346,6 +351,51 @@ static void applyFlipOverAfterCrashModeToMotors(void)
         }
     }
 }
+#ifdef USE_RPM_LIMITER
+
+static bool isMotorSaturated(void) //Placeholder function
+{
+    for (int i = 0; i < getMotorCount(); i++) {
+        if (motor[i] >= motorConfig()->maxthrottle) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static void applyRPMLimiter(mixerRuntime_t *mixer)
+{
+    if (RPM_LIMIT_ACTIVE && motorConfig()->dev.useDshotTelemetry && ARMING_FLAG(ARMED)) {
+        float averageRPM = getDshotAverageRpm() / 10.0f;
+        float averageRPMSmoothed = pt1FilterApply(&mixer->averageRPMFilter, averageRPM);
+        float smoothedRPMError = averageRPMSmoothed - mixer->rpmLimiterRPMLimit;
+        // PID
+        float rpmLimiterP = smoothedRPMError * mixer->rpmLimiterPGain;
+        mixer->rpmLimiterI += smoothedRPMError * mixer->rpmLimiterIGain;
+        mixer->rpmLimiterI = MAX(0.0f, mixer->rpmLimiterI);
+        float rpmDerivative = smoothedRPMError - mixer->rpmLimiterPreviousSmoothedRPMError;
+        float rpmLimiterD = rpmDerivative * mixer->rpmLimiterDGain;
+        float pidOutput = rpmLimiterP + mixer->rpmLimiterI + rpmLimiterD;
+        // Throttle limit learning
+        if (smoothedRPMError > -10.0f && rcCommand[THROTTLE] < rxConfig()->maxcheck) {
+            mixer->rpmLimiterExpectedThrottleLimit *= 1.0f - 4.8f / pidGetPidFrequency();
+        } else if (pidOutput < -0.05f && rcCommand[THROTTLE] > rxConfig()->maxcheck && !isMotorSaturated()) { // Throttle accel corresponds with motor accel
+            mixer->rpmLimiterExpectedThrottleLimit *= 1.0f + 3.2f / pidGetPidFrequency();
+        }
+        mixer->rpmLimiterExpectedThrottleLimit = constrainf(mixer->rpmLimiterExpectedThrottleLimit, 0.01f, 1.0f);
+        throttle *= mixer->rpmLimiterExpectedThrottleLimit;
+        // Output
+        pidOutput = MAX(0.0f, pidOutput);
+        throttle = constrainf(throttle-pidOutput, 0.0f, 1.0f);
+        mixer->rpmLimiterPreviousSmoothedRPMError = smoothedRPMError;
+        DEBUG_SET(DEBUG_RPM_LIMITER, 0, smoothedRPMError);
+        DEBUG_SET(DEBUG_RPM_LIMITER, 1, throttle * 100.0f);
+        DEBUG_SET(DEBUG_RPM_LIMITER, 2, mixer->rpmLimiterI * 100.0f);
+        DEBUG_SET(DEBUG_RPM_LIMITER, 3, rpmLimiterD * 100.0f);
+    }
```

Comment:
I tried to refactor this a bit to make it easier to read. See the comment where I left a note for the D-term. You can remove the comment. It's just there so you are aware I made this one change to the D-term calculation.

The if-statement is further down now at around line 630 so it's easier to see on first glance in which cases the RPM limiter gets applied.

```suggestion
        static float prevError = 0.0f;
        static float i = 0.0f;

        const float averageRpm = pt1FilterApply(&mixer->averageRpmFilter, getDshotAverageRpm() / 10.0f);
        const float error = averageRpm - mixer->rpmLimiterRpmLimit;

        // PID
        const float p = error * mixer->rpmLimiterPGain;
        const float d = (error - prevError) * pidGetPidFrequency() * mixer->rpmLimiterDGain;  // <=== Notice that we need to adjust for different PID loop rates here. Otherwise the D-term needs a different value depending on the looprate.
        i += error * mixer->rpmLimiterIGain;
        i = MAX(0.0f, i);
        float pidOutput = p + i + d;

        // Throttle limit learning
        if (error > -10.0f && rcCommand[THROTTLE] < rxConfig()->maxcheck) {
            mixer->rpmLimiterExpectedThrottleLimit *= 1.0f - 4.8f / pidGetPidFrequency();
        } else if (pidOutput < -0.05f && rcCommand[THROTTLE] > rxConfig()->maxcheck && !isMotorSaturated()) { // Throttle accel corresponds with motor accel
            mixer->rpmLimiterExpectedThrottleLimit *= 1.0f + 3.2f / pidGetPidFrequency();
        }
        mixer->rpmLimiterExpectedThrottleLimit = constrainf(mixer->rpmLimiterExpectedThrottleLimit, 0.01f, 1.0f);
        throttle *= mixer->rpmLimiterExpectedThrottleLimit;
        
        // Output
        pidOutput = MAX(0.0f, pidOutput);
        throttle = constrainf(throttle - pidOutput, 0.0f, 1.0f);
        prevError = error;

        DEBUG_SET(DEBUG_RPM_LIMITER, 0, error);
        DEBUG_SET(DEBUG_RPM_LIMITER, 1, throttle * 100.0f);
        DEBUG_SET(DEBUG_RPM_LIMITER, 2, i * 100.0f);
        DEBUG_SET(DEBUG_RPM_LIMITER, 3, d * 100.0f);
```

Reply:
done, but pidGetPidFrequency adjustments for D left in mixer_init.c